File: src/main/java/com/fengsheng/card/ChengQing.java
package com.fengsheng.card;

import com.fengsheng.*;
import com.fengsheng.phase.MainPhaseIdle;
import com.fengsheng.phase.OnUseCard;
import com.fengsheng.phase.WaitForChengQing;
import com.fengsheng.protos.Common;
import com.fengsheng.protos.Fengsheng;
import org.apache.log4j.Logger;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;

public class ChengQing extends AbstractCard {
    private static final Logger log = Logger.getLogger(ChengQing.class);

    public ChengQing(int id, Common.color[] colors, Common.direction direction, boolean lockable) {
        super(id, colors, direction, lockable);
    }

    public ChengQing(int id, AbstractCard card) {
        super(id, card);
    }

    @Override
    public Common.card_type getType() {
        return Common.card_type.Cheng_Qing;
    }

    @Override
    public boolean canUse(Game g, Player r, Object... args) {
        Player target = (Player) args[0];
        int targetCardId = (Integer) args[1];
        Fsm fsm = g.getFsm();
        if (fsm instanceof MainPhaseIdle) {
            if (r != ((MainPhaseIdle) fsm).player()) {
                log.error("澄清的使用时机不对");
                return false;
            }
        } else if (fsm instanceof WaitForChengQing) {
            if (r != ((WaitForChengQing) fsm).askWhom) {
                log.error("澄清的使用时机不对");
                return false;
            }
        } else {
            log.error("澄清的使用时机不对");
            return false;
        }
        if (!target.isAlive()) {
            log.error("目标已死亡");
            return false;
        }
        Card targetCard = target.findMessageCard(targetCardId);
        if (targetCard == null) {
            log.error("没有这张情报");
            return false;
        }
        if (!targetCard.getColors().contains(Common.color.Black)) {
            log.error("澄清只能对黑情报使用");
            return false;
        }
        return true;
    }

    @Override
    public void execute(final Game g, final Player r, Object... args) {
        final Player target = (Player) args[0];
        final int targetCardId = (Integer) args[1];
        log.info(r + "对" + target + "使用了" + this);
        r.deleteCard(this.id);
        final Fsm fsm = g.getFsm();
        Fsm resolveFunc = () -> {
            Card targetCard = target.deleteMessageCard(targetCardId);
            log.info(target + "面前的" + targetCard + "被置入弃牌堆");
            g.getDeck().discard(targetCard);
            for (Player player : g.getPlayers()) {
                if (player instanceof HumanPlayer p) {
                    p.send(Fengsheng.use_cheng_qing_toc.newBuilder()
                            .setCard(this.toPbCard()).setPlayerId(p.getAlternativeLocation(r.location()))
                            .setTargetPlayerId(p.getAlternativeLocation(target.location())).setTargetCardId(targetCardId).build());
                }
            }
            g.getDeck().discard(this);
            return new ResolveResult(fsm, true);
        };
        if (fsm instanceof MainPhaseIdle)
            g.resolve(new OnUseCard(((MainPhaseIdle) fsm).player(), r, this, r, resolveFunc));
        else if (fsm instanceof WaitForChengQing)
            g.resolve(new OnUseCard(((WaitForChengQing) fsm).whoseTurn, r, this, r, resolveFunc));
    }

    @Override
    public String toString() {
        return Card.cardColorToString(colors) + "澄清";
    }

    public static boolean ai(MainPhaseIdle e, Card card) {
        Player player = e.player();
        List<PlayerAndCard> playerAndCards = new ArrayList<>();
        var identity = player.getIdentity();
        for (Player p : player.getGame().getPlayers()) {
            if ((p.equals(player) || identity != Common.color.Black && identity == p.getIdentity()) && p.isAlive()) {
                for (Card c : p.getMessageCards().values()) {
                    if (c.getColors().contains(Common.color.Black))
                        playerAndCards.add(new PlayerAndCard(p, c));
                }
            }
        }
        if (playerAndCards.isEmpty()) return false;
        PlayerAndCard p = playerAndCards.get(ThreadLocalRandom.current().nextInt(playerAndCards.size()));
        GameExecutor.post(player.getGame(), () -> card.execute(player.getGame(), player, p.player(), p.card().getId()), 2, TimeUnit.SECONDS);
        return true;
    }

    private record PlayerAndCard(Player player, Card card) {

    }
}

File: src/main/kotlin/card/ChengQing.kt
package com.fengsheng.card

import com.fengsheng.*
import com.fengsheng.phase.*
import com.fengsheng.protos.Common.card_type.Cheng_Qing
import com.fengsheng.protos.Common.color
import com.fengsheng.protos.Common.color.Black
import com.fengsheng.protos.Common.direction
import com.fengsheng.protos.Common.secret_task.*
import com.fengsheng.protos.useChengQingToc
import com.fengsheng.skill.ConvertCardSkill
import com.fengsheng.skill.SkillId.*
import com.fengsheng.skill.cannotPlayCard
import org.apache.logging.log4j.kotlin.logger
import java.util.concurrent.TimeUnit

class ChengQing : Card {
    constructor(id: Int, colors: List<color>, direction: direction, lockable: Boolean) :
        super(id, colors, direction, lockable)

    constructor(id: Int, card: Card) : super(id, card)

    /**
     * 仅用于“作为澄清使用”
     */
    internal constructor(originCard: Card) : super(originCard)

    override val type = Cheng_Qing

    override fun canUse(g: Game, r: Player, vararg args: Any): Boolean {
        if (r.cannotPlayCard(type)) {
            logger.error("你被禁止使用澄清")
            r.sendErrorMessage("你被禁止使用澄清")
            return false
        }
        val target = args[0] as Player
        val targetCardId = args[1] as Int
        val fsm = g.fsm
        if (fsm is MainPhaseIdle) {
            if (r !== fsm.whoseTurn) {
                logger.error("澄清的使用时机不对")
                r.sendErrorMessage("澄清的使用时机不对")
                return false
            }
        } else if (fsm is WaitForChengQing) {
            if (r !== fsm.askWhom) {
                logger.error("澄清的使用时机不对")
                r.sendErrorMessage("澄清的使用时机不对")
                return false
            }
            if (target !== fsm.whoDie) {
                logger.error("正在求澄清的人是${fsm.whoDie}")
                r.sendErrorMessage("正在求澄清的人是${fsm.whoDie}")
                return false
            }
        } else {
            logger.error("澄清的使用时机不对")
            r.sendErrorMessage("澄清的使用时机不对")
            return false
        }
        if (!target.alive) {
            logger.error("目标已死亡")
            r.sendErrorMessage("目标已死亡")
            return false
        }
        val targetCard = target.messageCards.find { c -> c.id == targetCardId }
        if (targetCard == null) {
            logger.error("没有这张情报")
            r.sendErrorMessage("没有这张情报")
            return false
        }
        if (!targetCard.isBlack()) {
            logger.error("澄清只能对黑情报使用")
            r.sendErrorMessage("澄清只能对黑情报使用")
            return false
        }
        return true
    }

    override fun execute(g: Game, r: Player, vararg args: Any) {
        val target = args[0] as Player
        val targetCardId = args[1] as Int
        logger.info("${r}对${target}使用了$this")
        r.deleteCard(id)
        val fsm = g.fsm as ProcessFsm
        val resolveFunc = { _: Boolean ->
            val targetCard = target.deleteMessageCard(targetCardId)!!
            logger.info("${target}面前的${targetCard}被置入弃牌堆")
            g.deck.discard(targetCard)
            g.players.send {
                useChengQingToc {
                    card = toPbCard()
                    playerId = it.getAlternativeLocation(r.location)
                    targetPlayerId = it.getAlternativeLocation(target.location)
                    this.targetCardId = targetCardId
                }
            }
            if (fsm is MainPhaseIdle) {
                OnFinishResolveCard(fsm.whoseTurn, r, target, getOriginCard(), Cheng_Qing, fsm)
            } else {
                val newFsm = UseChengQingOnDying(fsm as WaitForChengQing)
                OnFinishResolveCard(fsm.whoseTurn, r, target, getOriginCard(), Cheng_Qing, newFsm)
            }
        }
        g.resolve(ResolveCard(fsm.whoseTurn, r, target, getOriginCard(), Cheng_Qing, resolveFunc, fsm))
    }

    override fun toString(): String {
        return "${cardColorToString(colors)}澄清"
    }

    companion object {
        fun ai(e: MainPhaseIdle, card: Card, convertCardSkill: ConvertCardSkill?): Boolean {
            val player = e.whoseTurn
            !(player.identity == Black && player.secretTask in listOf(Killer, Pioneer, Sweeper)) || return false
            !player.cannotPlayCard(Cheng_Qing) || return false
            // 自己是秦无命和裴玲，如果还没有使用技能就不用澄清
            player.findSkill(PIN_MING_SAN_LANG) == null || player.getSkillUseCount(PIN_MING_SAN_LANG) > 0 || return false
            player.findSkill(JIAO_JI) == null || player.getSkillUseCount(JIAO_JI) > 0 || return false

            val g = player.game!!
            var value = 10
            var playerAndCard: PlayerAndCard? = null
            for (p in g.sortedFrom(g.players, player.location)) {
                p.alive && (p === player || !g.isEarly) && p.isPartnerOrSelf(player) || continue
                for (c in p.messageCards.toList()) {
                    c.isBlack() || continue
                    val v = player.calculateRemoveCardValue(player, p, c)
                    if (v > value) {
                        value = v
                        playerAndCard = PlayerAndCard(p, c)
                    }
                }
            }
            playerAndCard ?: return false
            GameExecutor.post(g, {
                convertCardSkill?.onConvert(player)
                card.asCard(Cheng_Qing).execute(g, player, playerAndCard.player, playerAndCard.card.id)
            }, 3, TimeUnit.SECONDS)
            return true
        }
    }
}
